Add a otelcol.exporter.loadbalancing component (#4148)
grafana/agent: docs/sources/flow/reference/components/otelcol.exporter.loadbalancing.md

---
title: otelcol.exporter.loadbalancing
labels:

>   stage: beta

Author: I labeled this as "beta" because the Otel Collector [also](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/loadbalancingexporter) labeled it as beta

`otelcol.exporter.loadbalancing` accepts traces from other `otelcol` components
and writes them over the network using the OpenTelemetry Protocol (OTLP) protocol. 

> **NOTE**: `otelcol.exporter.loadbalancing` is a wrapper over the upstream
> OpenTelemetry Collector `loadbalancing` exporter. Bug reports or feature requests will
> be redirected to the upstream repository, if necessary.

Multiple `otelcol.exporter.loadbalancing` components can be specified by giving them
different labels.

`otelcol.exporter.loadbalancing` is especially useful for backends configured with tail-based samplers
which make a decision based on the view of the full trace.

## Usage

```river
otelcol.exporter.loadbalancing "LABEL" {
  resolver {
    ...
  }
  protocol {
    otlp {
      client {}
    }
  }
}
```

## Arguments

`otelcol.exporter.loadbalancing` supports the following arguments:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`routing_key` | `string` | Routing strategy for load balancing. | `"traceID"` | no

The `routing_key` attribute determines how to route spans across endpoints. Its value could be one of the following:

> * `"service"`: exports spans based on their service name. This is useful when using processors like the span metrics, 

Author: We don't [yet](https://github.com/grafana/agent/issues/2295) have a span metrics processor :) But I added this comment because it was mentioned in the Collector docs and seems useful

retried.

{{< docs/shared lookup="flow/reference/components/otelcol-retry-block.md" source="agent" >}}

## Exported fields

The following fields are exported and can be referenced by other components:

Name | Type | Description
---- | ---- | -----------
`input` | `otelcol.Consumer` | A value that other components can use to send telemetry data to.

`input` accepts `otelcol.Consumer` data for any telemetry signal (metrics,
logs, or traces).

## Component health

`otelcol.exporter.loadbalancing` is only reported as unhealthy if given an invalid
configuration.

## Debug information

`otelcol.exporter.loadbalancing` does not expose any component-specific debug
information.

## Example

This example accepts OTLP traces over gRPC. It then sends them in a load-balanced 
way to "localhost:55690" or "localhost:55700" according to their trace ID.

```river
otelcol.receiver.otlp "default" {
    grpc {}
    output {
        traces  = [otelcol.exporter.loadbalancing.default.input]
    }
}

otelcol.exporter.loadbalancing "default" {
    resolver {
        static {
            hostnames = ["localhost:55690", "localhost:55700"]
        }
    }
    protocol {
        otlp {

>             client {}

Author: It's a bit weird how we have to add all these blocks just so `client {}` gets populated with defaults. IMO the user would not expect that. They would expect not having blocks would simply init them to the default values mentioned in the docs. Instead, they'd be zero-initialised. It feels a bit wrong to have to make the blocks mandatory just so they get their default values, and I might raise an issue about this so we can discuss it separately.

The problem is that if they get set to their default values, what should we set the mandatory values to? I guess the argument for not setting defaults is that mandatory attributes are not set either. But then probably nothing should be set and the whole block should be a nil.

---
title: otelcol.exporter.loadbalancing
labels:
  stage: beta
---

# otelcol.exporter.loadbalancing

{{< docs/shared lookup="flow/stability/beta.md" source="agent" >}}

`otelcol.exporter.loadbalancing` accepts logs and/or traces from other `otelcol` components
and writes them over the network using the OpenTelemetry Protocol (OTLP) protocol. 

> **NOTE**: `otelcol.exporter.loadbalancing` is a wrapper over the upstream
> OpenTelemetry Collector `loadbalancing` exporter. Bug reports or feature requests will
> be redirected to the upstream repository, if necessary.

> Multiple `otelcol.exporter.loadbalancing` components can be specified by giving them

Author: This is actually a standard line which we put in a lot of docs pages. I'm not against changing it everywhere, but for the sake of consistency I'd prefer to leave it as it is in this page for now. Maybe it can be changed for all components in a separate PR.

> Multiple `otelcol.exporter.loadbalancing` components can be specified by giving them

Reviewer: Then lets leave it as-is for now - save it for a more general "cleanup" PR.

different labels.

Note that either a trace ID or a service name is used for the decision on which backend to use. 
The actual backend load isn't taken into consideration. Even though this load-balancer won't do 
round-robin balancing of the batches, the load distribution should be very similar among backends 
with a standard deviation under 5% at the current configuration.

`otelcol.exporter.loadbalancing` is especially useful for backends configured with tail-based samplers
which make a decision based on the view of the full trace.

> When a list of backends is updated, around `1/n` of the space will be changed, so that the same trace ID 

Reviewer 2: Agree with Clayton here, I am not sure what this means. What's a space? I think I can infer what it means but more text here would be nice.

Author: I agree that it is confusing. I just copy pasted it from the original Otel docs because tbh I'm not entirely sure about their math. I rephrased it differently, in a way that makes more sense to me, and if you think it makes sense then I'll also open a PR upstream to change the Otel docs. I changed their formula a little bit too.

might be directed to a different backend, where `n` is the number of backends. This should be stable enough 

> for most cases, and the higher the number of backends, the less disruption it should cause.

Author: The otel [docs](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/v0.63.0/exporter/loadbalancingexporter) also mention this:

> Still, if routing stability is important for your use case and your list of backends are constantly changing, consider using the groupbytrace processor. This way, traces are dispatched atomically to this exporter, and the same decision about the backend is made for the trace as a whole.

However I don't want to include this, because we don't have a groupbytrace processor and it'd be confusing to mention it imo.

> for most cases, and the higher the number of backends, the less disruption it should cause.

Reviewer: I'm a little fuzzy on what we are trying to say here... 

I'm still not sure about the wording (still passive) but... is this what you're stying?

> When a list of backends is updated, around `1/n` of the space will be changed, where `n` is the number of backends. After the update, it's possible that a trace ID could be directed to a different backend. 

And... what disruption will this cause. What's the impact? Do we need to consider and document the impact of trace IDs moving around?

## Usage

```river
otelcol.exporter.loadbalancing "LABEL" {
  resolver {
    ...
  }
  protocol {
    otlp {
      client {}
    }
  }
}
```

## Arguments

`otelcol.exporter.loadbalancing` supports the following arguments:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`routing_key` | `string` | Routing strategy for load balancing. | `"traceID"` | no

The `routing_key` attribute determines how to route signals across endpoints. Its value could be one of the following:
* `"service"`: exports spans based on their service name. This is useful when using processors like the span metrics, 
so all spans for each service are sent to consistent collector instances for metric collection. 
Otherwise, metrics for the same services would be sent to different collectors, making aggregations inaccurate.

> * `"traceID"`: exports spans based on their traceID.

Reviewer 2: Can we get more detail on how this is handled? Or a link to otel docs?

## Blocks

> The following blocks are supported inside the definition of

Reviewer: ```suggestion
The following blocks are supported in the definition of
```

Author: This is also a standard line which we put in many docs pages. If we are to change it, I think it should be done for all pages in a separate PR.

Reviewer: @ptodev No problem. Leave it as-is then. Better we opt for consistency in the phrasing now and deal with the bigger issues :-)

The `>` symbol indicates deeper levels of nesting. For example, `resolver > static`
refers to a `static` block defined inside a `resolver` block.

[resolver]: #resolver-block
[static]: #static-block
[dns]: #dns-block
[protocol]: #protocol-block
[otlp]: #otlp-block
[client]: #client-block
[tls]: #tls-block
[keepalive]: #keepalive-block
[queue]: #queue-block
[retry]: #retry-block

### resolver block

The `resolver` block configures how to retrieve the endpoint to which this exporter will send data.

Inside the `resolver` block, either the [dns][] block or the [static][] block 
should be specified. If both `dns` and `static` are specified, `dns` takes precedence.

### static block

The `static` block configures a list of endpoints which this exporter will send data to.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`hostnames` | `list(string)` | List of endpoints to export to. |  | yes

### dns block

The `dns` block periodically resolves an IP address via the DNS `hostname` attribute. This IP address 
and the port specified via the `port` attribute will then be used by the gRPC exporter 

> as the endpoint to which to send data to.

Reviewer 2: There might be some inconsistent verbiage, `export to` and `send to`.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`hostname` | `string`   | DNS hostname to resolve. |  | yes

> `port`     | `string`   | Port to be used with the resolved IP address. | `"4317"` | no

Reviewer 2: Should this be expanded out, because at first glance it looks like the hostname port. `Port to be used with the IP addresses returned from the DNS hostname`?

`interval` | `duration` | Resolver interval. | `"5s"` | no
`timeout`  | `duration` | Resolver timeout. | `"1s"`  | no

### protocol block

The `protocol` block configures protocol-related settings for exporting.
At the moment only the OTLP protocol is supported.

### otlp block

The `otlp` block configures OTLP-related settings for exporting.

### client block

The `client` block configures the gRPC client used by the component. 
The endpoints used by the client block are the ones from the `resolver` block

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`compression` | `string` | Compression mechanism to use for requests. | `"gzip"` | no
`read_buffer_size` | `string` | Size of the read buffer the gRPC client to use for reading server responses. | | no
`write_buffer_size` | `string` | Size of the write buffer the gRPC client to use for writing requests. | `"512KiB"` | no
`wait_for_ready` | `boolean` | Waits for gRPC connection to be in the `READY` state before sending data. | `false` | no
`headers` | `map(string)` | Additional headers to send with the request. | `{}` | no

> `balancer_name` | `string` | Which gRPC client-side load balancer to use for requests. | | no

Reviewer 2: Isn't the default here `pick_first`?

Reviewer 2: This still feels a bit odd, the functional default is `pick_first` . @clayton-cornell for his thoughts on it. 
`When unspecified, `pick_first` is used.` -> This feels like the definition of a default when applied to a single field.

`auth` | `capsule(otelcol.Handler)` | Handler from an `otelcol.auth` component to use for authenticating requests. | | no

{{< docs/shared lookup="flow/reference/components/otelcol-compression-field.md" source="agent" >}}

The `balancer_name` argument controls what client-side load balancing mechanism
to use. See the gRPC documentation on [Load balancing][] for more information.
When unspecified, `pick_first` is used.

> An HTTP proxy can be configured through the following environment variables:

Reviewer: ```suggestion
You can configure an HTTP proxy with the following environment variables:
```

The `keepalive` block configures keepalive settings for gRPC client
connections.

The following arguments are supported:

Name | Type | Description | Default | Required
---- | ---- | ----------- | ------- | --------
`ping_wait` | `duration` | How often to ping the server after no activity. | | no
`ping_response_timeout` | `duration` | Time to wait before closing inactive connections if the server does not respond to a ping. | | no
`ping_without_stream` | `boolean` | Send pings even if there is no active stream request. | | no

### queue block

The `queue` block configures an in-memory buffer of batches before data is sent
to the gRPC server.

{{< docs/shared lookup="flow/reference/components/otelcol-queue-block.md" source="agent" >}}

### retry block

The `retry` block configures how failed requests to the gRPC server are
retried.

{{< docs/shared lookup="flow/reference/components/otelcol-retry-block.md" source="agent" >}}

## Exported fields

The following fields are exported and can be referenced by other components:

Name | Type | Description
---- | ---- | -----------
`input` | `otelcol.Consumer` | A value that other components can use to send telemetry data to.

> `input` accepts `otelcol.Consumer` data for telemetry signals of these types:

Author: This way of presenting the exported fields is not standard. I customised it to highlight that only logs and traces work. Maybe I also mention unsupported types like this:

> These signal types are not supported:
> * metrics

We also have #4151 for this presentation issue

> * logs

Reviewer 2: Specify otel logs, since we also support loki?
